File: src/models/GroupingRule.js
import _ from 'underscore';

const storeData = [
  { id: 'PresetGroupingRule_1', title: '単一グループ', filename: 'sample1', params: '' },
  { id: 'PresetGroupingRule_2', title: 'ランダムグループ分け', filename: 'sample2', params: '{\n  "groupNum": 3,\n}', active: true },
];

class GroupingRule {
  constructor(id, title, filename, params, active) {
    this.id = id;
    this.title = title;
    this.filename = filename;
    this.params = params;
    this.active = active;
  }

  save() {
    const id = this.id;
    const json = storeData.find(data => id === data.id);
    if (json) {
      json.title = this.title;
      json.filename = this.filename;
      json.params = this.params;
      json.active = this.active;
    } else {
      storeData.push({
        id: this.id,
        title: this.title,
        filename: this.filename,
        params: this.params,
        active: this.active,
      });
    }
  }

  static of(title, filename, params) {
    const id = _.uniqueId('RecommendRule_');
    return new GroupingRule(id, title, filename, params);
  }

  static findOne(id) {
    const json = storeData.find(data => id === data.id);
    return json ? new GroupingRule(json.id, json.title, json.filename, json.params) : null;
  }

  static findAll() {
    return storeData.map((data) => {
      return new GroupingRule(data.id, data.title, data.filename, data.params);
    });
  }
}

export default GroupingRule;

File: repositories/groupingRule.py
import importlib
import json
import utils.idGenerator as idGenerator
import repositories.account as account

storeData = [
  { 'id': 'PresetGroupingRule_1', 'title': '単一グループ', 'filename': 'singleGrouping', 'params': '' },
  { 'id': 'PresetGroupingRule_2', 'title': 'ランダムグループ分け', 'filename': 'randamGrouping', 'params': '{\n  "groupNum": 2\n}', 'active': True },
]

def save(groupingRule):
    storedData = findOne(groupingRule['id'])
    if storedData is not None :
        storedData['title'] = groupingRule.get('title', None)
        storedData['filename'] = groupingRule.get('filename', None)
        storedData['params'] = groupingRule.get('params', None)
        storedData['active'] = groupingRule.get('active', False)
    else :
        storeData.append({
            'id': groupingRule.get('id', None),
            'title': groupingRule.get('title', None),
            'filename': groupingRule.get('filename', None),
            'params': groupingRule.get('params', None),
            'active': groupingRule.get('active', False) })

def findOne(id):
    for data in storeData:
        if id == data['id']:
            return data
    return None

def findAll():
    return storeData

def findActiveRule():
    for data in storeData:
        if data.get('active', False):
            return data

def trialGrouping(groupingRule):
    m = importlib.import_module('rules.group.' + groupingRule["filename"])
    params = json.loads(groupingRule.get("params", {}))
    groupsMembers = m.execute(account.findAllAccounts(), params)
    return list(map(lambda members: {'id': idGenerator.generate('Group_'), 'members': members}, groupsMembers))
